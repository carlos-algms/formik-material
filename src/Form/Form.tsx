import { Formik, FormikHelpers, FormikProps } from 'formik';
import React, { ReactNode } from 'react';
import type PropsBase from '../shared/PropsBase';

export type OnSubmitFormCallback<T = any, R = any> = {
  (values: T, formikActions: FormikHelpers<T>): R;
  (values: T, formikActions: FormikHelpers<T>): Promise<R>;
};

export type FormProps<T> = PropsBase & {
  initialValues?: T;
  onSubmitForm: OnSubmitFormCallback<T>;
  validationSchema?: any | (() => any);
  children?: ReactNode;
};

/**
 * Form Component
 */
export default function Form({
  children,
  initialValues,
  onSubmitForm,
  validationSchema,
  className,
}: FormProps<any>) {
  if (!initialValues) {
    if (process.env.NODE_ENV !== 'production') {
      console.warn('No initial values provided to the <Form> Component');
    }
    return null;
  }

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={onSubmitForm}
      validationSchema={validationSchema}
    >
      {({ handleSubmit }: FormikProps<any>) => {
        return (
          <form onSubmit={handleSubmit} noValidate autoComplete="off" className={className}>
            {children}
          </form>
        );
      }}
    </Formik>
  );
}
