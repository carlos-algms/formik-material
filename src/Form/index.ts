export { default } from './Form';
export type { FormProps, OnSubmitFormCallback } from './Form';
