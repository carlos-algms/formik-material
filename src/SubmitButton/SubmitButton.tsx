import Button from '@material-ui/core/Button';
import { useFormikContext } from 'formik';
import React, { memo } from 'react';
import type PropsBase from '../shared/PropsBase';

function SubmitButton({ children, className }: PropsBase) {
	const { isSubmitting } = useFormikContext();

	return (
		<div className={className}>
			<Button variant="contained" color="primary" type="submit" disabled={isSubmitting}>
				{children}
			</Button>
		</div>
	);
}

export default memo(SubmitButton);
