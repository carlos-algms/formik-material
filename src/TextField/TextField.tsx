import MUITextField, { TextFieldProps } from '@mui/material/TextField';
import { Field, FieldProps } from 'formik';
import ErrorLabel from '../ErrorLabel';

const identity = (a: any) => a;

let idCount = 0;

const TextField = ({
  name = `TextField-${idCount++}`,
  className,
  fullWidth = true,
  onBlur = identity,
  onChange = identity,
  ...textFieldProps
}: TextFieldProps) => {
  return (
    <Field name={name}>
      {({ field, meta: { error, touched } }: FieldProps) => {
        const isErrorState = Boolean(error && touched);
        const { onBlur: formikOnBlur, onChange: formikOnChange, ...formikFieldProps } = field;

        return (
          <div className={className}>
            <MUITextField
              {...textFieldProps}
              {...formikFieldProps}
              onBlur={(e) => {
                formikOnBlur(e);
                onBlur(e);
              }}
              onChange={(e) => {
                formikOnChange(e);
                onChange(e);
              }}
              fullWidth={fullWidth}
              error={isErrorState}
            />
            {isErrorState && <ErrorLabel error={error} />}
          </div>
        );
      }}
    </Field>
  );
};

export default TextField;
