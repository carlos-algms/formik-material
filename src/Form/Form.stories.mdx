import { action } from '@storybook/addon-actions';
import { Meta, Canvas, Story, Props } from '@storybook/addon-docs/blocks';
import Form from './Form';
import SubmitButton from '../SubmitButton';
import useFormSubmittedValues from '../useFormSubmittedValues';
import TextField from '../TextField';

export const initialValues = { name: 'initial name' };

<Meta title="Form Component" component={Form} />

## Basic form implementation

<Canvas>
  <Story name="basic">
    {() => {
      const handle = useFormSubmittedValues(action('submitted'));
      return (
        <Form onSubmitForm={handle} initialValues={initialValues}>
          <TextField name="name" label="Name: " />
          <p>
            <SubmitButton>Submit</SubmitButton>
          </p>
        </Form>
      );
    }}
  </Story>
</Canvas>

## Forms should be able to render child functions

<Canvas>
  <Story name="render child function">
    {() => {
      const handle = useFormSubmittedValues(action('submitted'));
      return (
        <Form onSubmitForm={handle} initialValues={initialValues}>
          {() => (
            <>
              <TextField name="name" label="Name: " />
              <p>
                <SubmitButton>Submit</SubmitButton>
              </p>
            </>
          )}
        </Form>
      );
    }}
  </Story>
</Canvas>
