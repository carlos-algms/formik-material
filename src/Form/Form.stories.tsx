import { action } from '@storybook/addon-actions';
import { Meta, StoryObj } from '@storybook/react';
import SubmitButton from '../SubmitButton';
import TextField from '../TextField';
import useFormSubmittedValues from '../useFormSubmittedValues';
import Form from './Form';

const initialValues = { name: 'initial name' };

const meta: Meta<typeof Form> = {
  title: 'Form Component',
  component: Form,
};

export default meta;

type Story = StoryObj<typeof Form>;

export const Basic: Story = {
  render: () => {
    const handle = useFormSubmittedValues(action('submitted'));
    return (
      <Form onSubmitForm={handle} initialValues={initialValues}>
        <TextField name="name" label="Name: " />
        <p>
          <SubmitButton>Submit</SubmitButton>
        </p>
      </Form>
    );
  },
};

export const ChildFn: Story = {
  name: 'Render children as function',
  render: () => {
    const handle = useFormSubmittedValues(action('submitted'));
    return (
      <Form onSubmitForm={handle} initialValues={initialValues}>
        {() => (
          <>
            <TextField name="name" label="Name: " />
            <p>
              <SubmitButton>Submit</SubmitButton>
            </p>
          </>
        )}
      </Form>
    );
  },
};
