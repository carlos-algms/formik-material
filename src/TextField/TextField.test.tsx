import { act, fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import Form from '../Form';
import SubmitButton from '../SubmitButton';
import { delay } from '../tests/helpers';
import TextField from './TextField';

test('should render a MUI field and error label', () => {
  const submitMock = jest.fn();
  const values = { firstName: 'John' };

  const { container } = render(
    <Form onSubmitForm={submitMock} initialValues={values} initialError>
      <TextField name="firstName" className="wrapper-class" />
    </Form>,
  );

  const wrapper = container.querySelector('.wrapper-class');

  expect(wrapper).toMatchSnapshot();
});

test('should render error label', async () => {
  const submitMock = jest.fn();
  const validateMock = jest.fn(() => ({
    firstName: 'Not valid',
  }));

  const values = { firstName: 'Jane' };

  const { container } = render(
    <Form validate={validateMock} onSubmitForm={submitMock} initialValues={values} initialError>
      <TextField name="firstName" className="wrapper-class" />
      <SubmitButton>Send</SubmitButton>
    </Form>,
  );

  const submitButton = screen.getByRole('button');

  await act(async () => {
    fireEvent.click(submitButton);
    await delay();
  });

  const wrapper = container.querySelector('.wrapper-class');

  expect(wrapper).toMatchSnapshot();
});
