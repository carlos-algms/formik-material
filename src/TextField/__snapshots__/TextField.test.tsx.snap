// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a MUI field and error label 1`] = `
<div
  class="wrapper-class"
>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input"
        name="firstName"
        type="text"
        value="John"
      />
    </div>
  </div>
</div>
`;

exports[`should render error label 1`] = `
<div
  class="wrapper-class"
>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline Mui-error Mui-error MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-invalid="true"
        class="MuiInputBase-input MuiInput-input"
        name="firstName"
        type="text"
        value="Jane"
      />
    </div>
  </div>
  <div
    class="MuiTypography-root error-label makeStyles-error-1 MuiTypography-body2"
  >
    Not valid
  </div>
</div>
`;
